src/components/ParserComponent.tsx

```diff
@@ -143,81 +144,95 @@ export default class ParserComponent extends React.Component<ParserInterface,
-                  if (val.isValid()) {
-                    b[keyInt] = val.valueOf();
-                  } else {
-                    b[keyInt] = -1;
-                  }
-                }
-                return (a[keyInt] - b[keyInt]);
-              });
-
-              this.setState(() => {
-                return {
-                  data,
-                };
-              });
-            }
-          }
-        }
-      }
-      if (doneTheWork) {
-        // state should be updated
-        assert.notStrictEqual(this.state.data, [],
-            'sortData(): this.state.data is empty (not updated)');
         return true;
       } else {
-        throw new Error('The file uploaded has no dates.');
+        throw new Error('The file uploaded has no data.');
       }
     }
 
+    /**
+     * purpose: looks for the column that has a valid date in the first row
+     * @precondition a csv has been uploaded and its data is stored in an array
+     * @param {Array} data: the array of data to be checked
+     * @return {any}: return the column name of the first row
+     * that contains the valid date of given date format
+   */
+    lookForDateKey(data: Array<object>): any {
+      assert.notStrictEqual(data[0], null,
+          'sortData(): data[0] is null');
+      for (const [key, value] of Object.entries(data[0])) {
+        const date1 = moment(String(value), this.state.formatString);
+        if (moment(date1, this.state.formatString).isValid()) {
+          return key;
+        }
+      }
+    };
+
+    /**
+   * Purpose: sorts the array of data
+   * @precondition data contains a column of valid numbers that
+     * can be interpreted as a date in the selected date format
+   * @postcondition the data stored in the array is sorted by some date column
+   * @param {Array} data: the array of data to sort
+     * @param {any} key: the column name by which dates must be sorted
+   * @return {boolean}: true if data is sorted correctly
+     * and false if it failed to sort
+   */
+    sortData(data: Array<object>, key:any): boolean {
+      const formatString = this.state.formatString;
+      const initialData = this.state.data;
+      const keyInt = `${key}_num`;
+
+      TimSort.sort(data, function(a: any, b: any) {
+        let val: any;
+        if (!a.hasOwnProperty(keyInt)) {
+          val = moment(a[key], formatString);
+          if (val.isValid()) {
+            a[keyInt] = val.valueOf();
+          } else {
+            a[keyInt] = -1;
+          }
+        }
+
+        if (!b.hasOwnProperty(keyInt)) {
+          val = moment(b[key], formatString);
+          if (val.isValid()) {
+            b[keyInt] = val.valueOf();
+          } else {
+            b[keyInt] = -1;
+          }
+        }
+        return (a[keyInt] - b[keyInt]);
+      });
+
+      this.setState(() => {
+        return {
+          data,
+        };
+      });
+      // state should be updated
+      assert.notStrictEqual(this.state.data, [],
```

Comment:
This assertion doesn't make much sense, it never was empty was it? Shouldn' t you be checking that it changed aka was sorted?